### PR TITLE
[1632] In the Form Engine: Add a property sheet plugin type for repository path that allow author to browse a path 

### PR DESCRIPTION
**Pull request by aachavarria**
https://github.com/craftercms/craftercms/issues/1632

### CODE REVIEW

File: static-assets/components/cstudio-forms/data-sources/shared-content.js [hunk 1]
```diff
@@ -367,8 +367,8 @@ YAHOO.extend(CStudioForms.Datasources.SharedContent, CStudioForms.CStudioFormDat
       {
         label: CMgs.format(langBundle, 'repositoryPath'),
         name: 'repoPath',
-        type: 'string',
-        defaultValue: '/site/',
+        type: 'repository',
```
Comment (rart) [inline static-assets/components/cstudio-forms/data-sources/shared-content.js]: let's call this `content-path-input`

File: static-assets/themes/cstudioTheme/css/console.css [hunk 2]
```diff
@@ -24,6 +24,25 @@ body, tbody {
   height: 100%;
 }
 
+.repository-selector {
```
Comment (rart) [inline static-assets/themes/cstudioTheme/css/console.css]: content-path-input

File: ui/app/src/components/Dialogs/PathBrowserDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+  ok: {
+    id: 'words.accept',
+    defaultMessage: 'Accept'
+  },
+  cancel: {
+    id: 'words.cancel',
+    defaultMessage: 'Cancel'
+  },
+  create: {
+    id: 'path.browser.create',
```
Comment (rart) [inline ui/app/src/components/Dialogs/PathBrowserDialog.tsx]: pathBrowser.createFolderButtonLabel

File: ui/app/src/components/Dialogs/PathBrowserDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,254 @@
+/*
```
Comment (rart) [inline ui/app/src/components/Dialogs/PathBrowserDialog.tsx]: Let's call this file/component PathSelectionDialog

File: ui/app/src/components/Dialogs/PathBrowserDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+  };
+
+  const onPathChanged = (path: string) => {
+    setCurrentPath(path);
+    setExpanded(getAllPaths(path, rootPath));
+  };
+
+  return (
+    <>
+      <DialogHeader title="Select Path" onDismiss={onClose} />
```
Comment (rart) [inline ui/app/src/components/Dialogs/PathBrowserDialog.tsx]: Select Path should be i18n.
Add prop for `title` and in here, make "Select Path" the default message if the prop is not supplied.

File: ui/app/src/components/Navigation/FolderBrowserTreeView.tsx [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+function RenderTreeNode({ node }: RenderTreeNode) {
+  const classes = useStyles({});
+  return node.id === 'loading' ? (
+    <div className={classes.loading}>
+      <Typography>
+        <FormattedMessage id="folderBrowserTreeView.loading" defaultMessage={'Loading...'} />
+      </Typography>
+      <CircularProgress size={16} />
```
Comment (rart) [inline ui/app/src/components/Navigation/FolderBrowserTreeView.tsx]: Looks like this is still inverted. The spinner should be to the left of the message.

Use \<FormattedMessage id=words.loading/>&hellip;

File: ui/app/src/components/Dialogs/PathBrowserDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+  };
+
+  const onNodeSelected = (event: React.ChangeEvent<{}>, nodeId: string) => {
+    setCurrentPath(nodeId);
+  };
+
+  const onPathChanged = (path: string) => {
+    setCurrentPath(path);
+    setExpanded(getAllPaths(path, rootPath));
+  };
```
Comment (rart) [inline ui/app/src/components/Dialogs/PathBrowserDialog.tsx]: While verifying a path, the accept button should be disabled.

When there's a manual path entry on the input, if the connection is slow enough, there's a moment when you could accept a path that does not exist.

File: ui/app/src/utils/path.ts [hunk 8]
```diff
@@ -105,6 +105,14 @@ export function getParentsFromPath(path: string, rootPath: string): string[] {
   return [rootPath, ...splitPath.map((value, i) => `${rootPath}/${splitPath.slice(1, i + 1).join('/')}`).splice(2)];
 }
 
+export function getAllPaths(path: string, rootPath: string) {
```
Comment (rart) [inline ui/app/src/utils/path.ts]: getIndividualPaths

File: static-assets/components/cstudio-forms/data-sources/shared-content.js [hunk 9]
```diff
@@ -367,8 +367,8 @@ YAHOO.extend(CStudioForms.Datasources.SharedContent, CStudioForms.CStudioFormDat
...
+        defaultValue: '/site',
```
Comment (rart) [inline static-assets/components/cstudio-forms/data-sources/shared-content.js]: make the default & root path `/site/` (with the slash at the end too)

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/legacy/src/components/cstudio-admin/mods/content-type-propsheet/repository.js [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+          className="repository-selector--input"
+          placeholder='/site/...'
+          onBlur={validations.startsWith ? onBlur : null}
+          onChange={(e) => onChange(e.target.value)}
+        />
+        <i
+          className="repository-selector--icon fa fa-search"
+          aria-hidden="true"
+          onClick={openPathBrowser}
+        />
```
Comment (rart) [inline ui/legacy/src/components/cstudio-admin/mods/content-type-propsheet/repository.js]: Wrap with a button. Make sure you can tab between input and button in that order.

File: ui/app/src/components/Navigation/FolderBrowserTreeView.tsx [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+  id: string;
+  name: string;
+  children?: TreeNode[];
+  fetched?: boolean;
+}
+
+const useStyles = makeStyles(() =>
+  createStyles({
+    wrapper: {
+      minHeight: '60vh',
```
Comment (rart) [inline ui/app/src/components/Navigation/FolderBrowserTreeView.tsx]: Move the min height to the dialog (body). And, let's use 40vh.

File: static-assets/components/cstudio-forms/data-sources/shared-content.js (hunk 1 above)
Comment (rart) [reply static-assets/components/cstudio-forms/data-sources/shared-content.js]: Have a look around for other inputs that would too benefit from using this selector. An example is the "Browse path" prop of data sources.

File: ui/app/src/components/Dialogs/PathSelectionDialog.tsx [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+    setCurrentPath(path);
+    setExpanded(getIndividualPaths(path, rootPath));
+  };
+
+  return (
+    <>
+      <DialogHeader
+        title={title
+          ? title
+          : <FormattedMessage id="path.dialog.title" defaultMessage="Select Path" />}
```
Comment (rart) [inline ui/app/src/components/Dialogs/PathSelectionDialog.tsx]: change message id to pathSelectionDialog.title

File: ui/app/src/utils/path.ts [hunk 13]
```diff
@@ -105,6 +105,14 @@ export function getParentsFromPath(path: string, rootPath: string): string[] {
   return [rootPath, ...splitPath.map((value, i) => `${rootPath}/${splitPath.slice(1, i + 1).join('/')}`).splice(2)];
 }
 
+export function getIndividualPaths(path: string, rootPath: string) {
```
Comment (rart) [inline ui/app/src/utils/path.ts]: why do you need the rootpath as a separate argument?

Comment (aachavarria) [reply ui/app/src/utils/path.ts]: I was using two different rootpath. Now only one, we can remove it, if you want to

**Review by rart — APPROVED** [review]

**Pull request merged**